distsys/resources/twopc.go

```diff
@@ -252,87 +360,228 @@ func (state CriticalSectionState) String() string {
 	return "unknown"
 }
 
-// Abort aborts the current critical section state. If this node has already
-// completed a 2PC precommit, then it should rollback the PreCommit.
-func (res *TwoPCArchetypeResource) Abort() chan struct{} {
+func (res *TwoPCArchetypeResource) startTiming(key string) {
+	if timersEnabled {
+		res.timerMutex.Lock()
+		res.timers[key] = time.Now()
+		res.timerMutex.Unlock()
+	}
+}
+
+func (res *TwoPCArchetypeResource) stopTiming(key string) {
+	if timersEnabled {
+		res.timerMutex.Lock()
+		delete(res.timers, key)
+		res.timerMutex.Unlock()
+	}
+}
+
+func (res *TwoPCArchetypeResource) enterMutex(name string) {
 	res.mutex.Lock()
-	res.log(
-		"Aborts 2PC State %s, CS State: %s, replicas %s",
-		res.twoPCState,
-		res.criticalSectionState,
-		res.replicas,
-	)
-	res.value = res.oldValue
-	if res.criticalSectionState == hasPreCommitted {
-		res.log("Rollback replicas due to Abort()")
-		for _, r := range res.replicas {
-			request := makeAbort(res.archetypeID)
+	res.startTiming(fmt.Sprintf("lock-%s", name))
+}
+
+func (res *TwoPCArchetypeResource) leaveMutex(name string) {
+	res.stopTiming(fmt.Sprintf("lock-%s", name))
+	res.mutex.Unlock()
+}
+
+func (res *TwoPCArchetypeResource) inMutex(name string, action func()) {
+	res.enterMutex(name)
+	action()
+	res.leaveMutex(name)
+}
+
+func checkTimers(res *TwoPCArchetypeResource) {
+	res.timerMutex.Lock()
+	for key, start := range res.timers {
+		duration := time.Now().Sub(start)
+		if duration > 3*time.Second {
+			res.warn("Timer %s: %s", key, duration)
+		}
+	}
+	res.timerMutex.Unlock()
+}
+
+func (res *TwoPCArchetypeResource) rollback() {
+	res.enterMutex("rollback")
+	request := res.makeAbort()
+	res.leaveMutex("rollback")
+	res.sendToMajority("Abort", func(i int, r ReplicaHandle, isDone func() bool) bool {
+		for !isDone() {
 			var reply TwoPCResponse
 			call := r.Send(request, &reply)
-			<-call
+			err := <-call
+			if err != nil {
+				res.log("Error during abort RPC to %i: %s", i, err)
+			} else if !reply.Accept {
+				res.log("%i rejected abort", i)
+			} else {
+				// OK
+				break
+			}
 		}
-	}
-	res.criticalSectionState = notInCriticalSection
-	res.mutex.Unlock()
+		return true
+	})
+}
+
+// Abort aborts the current critical section state. If this node has already
+// completed a 2PC precommit, then it should rollback the PreCommit.
+func (res *TwoPCArchetypeResource) Abort() chan struct{} {
+	res.inMutex("abort", func() {
+		res.value = res.oldValue
+		if res.criticalSectionState == hasPreCommitted {
+			res.log("Rollback replicas due to Abort()")
+			res.leaveMutex("abort")
+			res.rollback()
+			res.enterMutex("abort")
+		}
+		res.criticalSectionState = notInCriticalSection
+	})
 	return nil
 }
 
+type sendFunc func(int, ReplicaHandle, func() bool) bool
+
+func (res *TwoPCArchetypeResource) sendToMajority(name string, f sendFunc) bool {
```

Comment:
Actually I think we send to all replicas, receive from majority. Naming can be a bit misleading.